语法错误
python 解释器的语法检测不通过，必须在程序执行前就改正。
#语法错误示范一
if

#语法错误示范二
def test:
    pass

#语法错误示范三
print(haha

逻辑错误
#用户输入不完整(比如输入为空)或者输入非法(输入不是数字)
num=input(">>: ")
res1 = int(num)

异常
异常就是程序运行时发生错误的信号，然后程序异常退出。

int(s)
常见异常

异常类型	说明 ｜
属性错误	试图访问一个对象没有的属性，比如 foo.x，但是 foo 没有属性 x
IOErr或	输入/输出异常；基本上是无法打开文件
导入错误	无法引入模块或包；基本上是路径问题或名称错误
IndentationError （缩进错误）	语法错误（的子类） ；代码没有正确对齐
IndexError	下标索引超出序列边界，比如当x只有三个元素，却试图访问 x[5]
KeyError 错误	试图访问字典里不存在的键
KeyboardInterrupt （键盘中断）	Ctrl + C 被按下
名称错误	使用一个还未被赋予对象的变量
语法错误	Python 代码非法，代码不能编译
TypeError （类型错误）	传入对象类型与要求的不符合
UnboundLocalError 错误	试图访问一个还未被设置的局部变量，基本上是由于另有一个同名的全局变量，导致你以为正在访问它
ValueError	传入一个调用者不期望的值，即使值的类型是正确的

异常处理
python 解释器检测到错误，触发异常（也允许程序员自己触发异常）
程序员编写特定的代码，专门用来捕捉这个异常（这段代码与程序逻辑无关，与异常处理有关）
如果捕捉成功则进入另外一个处理分支，执行你为其定制的逻辑，使程序不会崩溃，这就是异常处理
异常是由程序的错误引起的，语法上的错误跟异常处理无关，必须在程序运行前就修正c

基本语法
try:
    被检测的代码块
except 异常类型：
    try 被检测的代码块，就执行这个位置的逻辑
...
else:
    正常执行逻辑
finally:
    扫尾工作

多分支加万能异常

s1 = 'hello'
try:
    int(s1)
except IndexError as e:
    print(e)
except KeyError as e:
    print(e)
except ValueError as e:
    print(e)
except Exception as e:#捕获所有异常
    print(e)

主动触发异常
try:
    raise TypeError('类型错误')
except Exception as e:
    print(e)
自定义异常
class EvaException(BaseException):
    def __init__(self,msg):
        self.msg=msg
    def __str__(self):
        return self.msg

try:
    raise EvaException('类型错误')
except EvaException as e:
    print(e)